src/inmanta/server/services/environment_metrics_service.py

```diff
@@ -206,27 +208,37 @@ async def _cleanup_old_metrics(self) -> None:
         environment setting.
         """
         async with Environment.get_connection() as con:
-            environments = await Environment.get_list(connection=con)
-            env_id_to_delete_older_than_timestamp: Dict[uuid.UUID, datetime] = {}
-            for env in environments:
-                if env.halted:
-                    continue
-                retention_time_in_hours: int = typing.cast(int, await env.get(ENVIRONMENT_METRICS_RETENTION))
-                if retention_time_in_hours <= 0:
-                    # Cleanups are disabled
-                    continue
-                delete_older_than: datetime = datetime.now().astimezone() - timedelta(hours=retention_time_in_hours)
-                env_id_to_delete_older_than_timestamp[env.id] = delete_older_than
-
-            if not env_id_to_delete_older_than_timestamp:
-                return
-            filter_condition = " OR ".join(
-                f"(environment=${(2*i)+1} AND timestamp < ${(2*i)+2})"
-                for i in range(len(env_id_to_delete_older_than_timestamp))
+            query = f"""
+            WITH env_and_retention_time_in_hours AS (
+                SELECT id, (CASE WHEN e.settings ? $1 THEN (e.settings->>$1)::integer ELSE $2 END) AS retention_time_in_hours
+                FROM {Environment.table_name()} AS e
+                WHERE e.halted IS FALSE
+            ), env_and_delete_before_timestamp AS (
+                SELECT e.id, ($3::timestamptz - make_interval(hours => e.retention_time_in_hours)) AS delete_before_timestamp
+                FROM env_and_retention_time_in_hours AS e
+                WHERE e.retention_time_in_hours > 0
+            ), delete_gauge AS (
+                DELETE FROM {EnvironmentMetricsGauge.table_name()} AS emg
+                WHERE EXISTS(
```

Comment:
I would think a `INNER JOIN` would have slightly better performance because the `EXISTS` will continue searching if it hits the right environment but the `AND` condition doesn't match.
That said, I don't think this will be a performance bottleneck, no need to change this on my account. I've just gotten in the habit of trying to analyze all queries I see a practice.